subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -0,0 +1,417 @@
+		zoap_next_block(&firmware_block_ctx);
+
+		SYS_LOG_DBG("total: %zd, current: %zd",
+			    firmware_block_ctx.total_size,
+			    firmware_block_ctx.current);
+
+		/* callback */
+		callback = lwm2m_firmware_get_write_cb();
+		if (callback) {
+			callback(0, payload, payload_len,
+				(firmware_block_ctx.current ==
+					firmware_block_ctx.total_size),
+				firmware_block_ctx.total_size);
+		}
+	}
+
+	/* TODO: Determine actual completion criteria */
+	if (firmware_block_ctx.current < firmware_block_ctx.total_size) {
+		token = zoap_header_get_token(check_response, &tkl);
+		ret = transfer_request(&firmware_block_ctx, token, tkl,
+				       do_firmware_transfer_reply_cb);
+	}
+
+	return ret;
+}
+
+static void firmware_transfer(struct k_work *work)
+{
+#if defined(CONFIG_NET_IPV6)
+	static struct sockaddr_in6 any_addr6 = { .sin6_addr = IN6ADDR_ANY_INIT,
+						.sin6_family = AF_INET6 };
+#endif
+#if defined(CONFIG_NET_IPV4)
+	static struct sockaddr_in any_addr4 = { .sin_addr = INADDR_ANY_INIT,
+					       .sin_family = AF_INET };
+#endif
+	struct net_if *iface;
+	int ret, port, family;
+
+	/* Server Peer IP information */
+	/* TODO: use parser on firmware_uri to determine IP version + port */
+	/* TODO: hard code IPv4 + port for now */
+	port = 5685;
+	family = AF_INET;
+
+#if defined(CONFIG_NET_IPV6)
+	if (family == AF_INET6) {
+		firmware_addr.family = family;
+		/* HACK: use firmware_uri directly as IP address */
+		net_addr_pton(firmware_addr.family, firmware_uri,
+			      &net_sin6(&firmware_addr)->sin6_addr);
+		net_sin6(&firmware_addr)->sin6_port = htons(5685);
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+	if (family == AF_INET) {
+		firmware_addr.family = family;
+		net_addr_pton(firmware_addr.family, firmware_uri,
+			      &net_sin(&firmware_addr)->sin_addr);
+		net_sin(&firmware_addr)->sin_port = htons(5685);
+	}
+#endif
+
+	ret = net_context_get(firmware_addr.family, SOCK_DGRAM, IPPROTO_UDP,
+			    &firmware_net_ctx);
+	if (ret) {
+		NET_ERR("Could not get an UDP context (err:%d)", ret);
+		return;
+	}
+
+	iface = net_if_get_default();
+	if (!iface) {
+		NET_ERR("Could not find default interface");
+		goto cleanup;
+	}
+
+#if defined(CONFIG_NET_IPV6)
+	if (firmware_addr.family == AF_INET6) {
+		ret = net_context_bind(firmware_net_ctx,
+				       (struct sockaddr *)&any_addr6,
+				       sizeof(any_addr6));
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+	if (firmware_addr.family == AF_INET) {
+		ret = net_context_bind(firmware_net_ctx,
+				       (struct sockaddr *)&any_addr4,
+				       sizeof(any_addr4));
+	}
+#endif
+
+	if (ret) {
+		NET_ERR("Could not bind the UDP context (err:%d)", ret);
+		goto cleanup;
+	}
+
+	SYS_LOG_DBG("Attached to port: %d", port);
+	ret = net_context_recv(firmware_net_ctx, firmware_udp_receive, 0, NULL);
+	if (ret) {
+		SYS_LOG_ERR("Could not set receive for net context (err:%d)",
+			    ret);
+		goto cleanup;
+	}
+
+	/* reset block transfer context */
+#if defined(NET_L2_BLUETOOTH)
```

Comment:
This looks suspicious, shouldn't this be CONFIG_NET_L2_BLUETOOTH?
And shouldn't we check this at runtime whether the interface we are using is BT or not? We support multiple interfaces, although I am not sure how well they work atm.